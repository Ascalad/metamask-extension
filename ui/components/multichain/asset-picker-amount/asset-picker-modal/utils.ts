// Default filter predicate for whether a token should be included in displayed list
export const shouldIncludeToken = (
  symbol: string,
  address?: string | null,
  tokenChainId?: string,
  debouncedSearchQuery: string,
  isMultiselectEnabled: boolean,
  selectedChainIds: string[],
  selectedChainId: string,
  filteredTokensAddresses: Set<string>,
  getTokenKey: (address?: string | null, tokenChainId?: string) => string,
) => {
  const trimmedSearchQuery = debouncedSearchQuery.trim().toLowerCase();
  const isMatchedBySearchQuery = Boolean(
    !trimmedSearchQuery ||
      symbol?.toLowerCase().indexOf(trimmedSearchQuery) !== -1 ||
      address?.toLowerCase().indexOf(trimmedSearchQuery) !== -1,
  );
  const isTokenInSelectedChain = isMultiselectEnabled
    ? tokenChainId && selectedChainIds?.indexOf(tokenChainId) !== -1
    : selectedNetwork?.chainId === tokenChainId;

  return Boolean(
    isTokenInSelectedChain &&
      isMatchedBySearchQuery &&
      !filteredTokensAddresses.has(getTokenKey(address, tokenChainId)),
  );
};
