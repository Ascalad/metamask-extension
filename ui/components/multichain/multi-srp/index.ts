export { SRPList } from './srp-list';
export { SelectSRP } from './select-srp';
export { ImportSRP } from './import-srp';
