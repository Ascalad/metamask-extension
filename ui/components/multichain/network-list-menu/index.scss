.multichain-network-list-menu-content-wrapper {
  &__dialog {
    height: 100vh;
    max-height: 100%;
  }
}

.multichain-network-list-menu {
  height: 100%;
  overflow: auto;
}
