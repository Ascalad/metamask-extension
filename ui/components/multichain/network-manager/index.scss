@use "design-system";


.network-manager__tab-list {
  display: flex;
  width: 100%;

  li {
    flex: 1;
    text-align: center;

    button {
      width: 100%;
      justify-content: center;
    }
  }
}


.network-manager__networks-button-icon {
  justify-content: center;
  display: flex;
}

.network-manager__additional-network-item {
  .mm-button-icon {
    --button-icon-size: 20px;
  }
}
