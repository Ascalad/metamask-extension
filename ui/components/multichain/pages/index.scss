@import 'page/';
@import 'send/';
@import 'connections/';
@import 'permissions-page/permissions-page';
@import 'sites-page/sites-page';
