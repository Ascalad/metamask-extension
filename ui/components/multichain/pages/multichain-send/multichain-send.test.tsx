import React from 'react';
import { act, RenderResult } from '@testing-library/react';
import { InternalAccount } from '@metamask/keyring-api';
import { createListenerMiddleware, configureStore } from '@reduxjs/toolkit';
import { AnyAction } from 'redux';
import {
  renderWithProvider,
  waitFor,
  fireEvent,
} from '../../../../../test/jest';
import mockMultichainSendState from '../../../../../test/data/mock-multichain-send-state-with-empty-draft.json';
import messages from '../../../../../app/_locales/en/messages.json';
import { getSelectedInternalAccount } from '../../../../selectors';
import {
  addNewDraft,
  editTransaction,
  FeeLevel,
  initialMultichainSendState,
  SendStage,
  updateSendAmount,
  updateStage,
} from '../../../../ducks/multichain-send/multichain-send';
import { getMultichainProviderConfig } from '../../../../selectors/multichain';
import rootReducer from '../../../../ducks';
import { CombinedBackgroundAndReduxState } from '../../../../store/store';
import { MockTransactionBuilder } from '../../../../ducks/multichain-send/transaction-builders/__mocks__/mock-transaction-builder';
import { MultichainNetworks } from '../../../../../shared/constants/multichain/networks';
import { AssetType } from '../../../../../shared/constants/transaction';
import { shortenAddress } from '../../../../helpers/utils/util';
import {
  DEFAULT_ROUTE,
  MULTICHAIN_CONFIRM_TRANSACTION_ROUTE,
} from '../../../../helpers/constants/routes';
import { MultichainSendPage } from './multichain-send';

const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useLocation: jest.fn(() => ({ search: '' })),
  useHistory: () => ({
    push: mockHistoryPush,
  }),
}));

jest.mock(
  '../../../../ducks/multichain-send/transaction-builders/transaction-builder.ts',
  () => {
    return {
      TransactionBuilderFactory: {
        getBuilder: jest.fn(() => {
          // @ts-expect-error mock doesn't need the constructor args
          return new MockTransactionBuilder();
        }),
      },
    };
  },
);

const mockTransactionId = 'test-id';

const baseStore = {
  ...mockMultichainSendState,
  metamask: {
    ...mockMultichainSendState.metamask,
    rates: {
      btc: {
        conversionDate: 1620710825.03,
        conversionRate: 56594.07,
        usdConversionRate: 56594.07,
      },
    },
    permissionHistory: {},
    pinnedAccountList: [],
    hiddenAccountList: [],
    balances: {
      'a39f7dd5-04f8-48a0-9359-6780d8fe397a': {
        'bip122:000000000933ea01ad0ee984209779ba/slip44:0': {
          amount: '0.00108995',
          unit: 'BTC',
        },
      },
    },
  },
};

const baseStoreWithDraft = {
  ...mockMultichainSendState,
  metamask: {
    ...mockMultichainSendState.metamask,
    rates: {
      btc: {
        conversionDate: 1620710825.03,
        conversionRate: 56594.07,
        usdConversionRate: 56594.07,
      },
    },
  },
  multichainSend: {
    stage: SendStage.DRAFT,
    draftTransactions: {
      [mockTransactionId]: {
        id: mockTransactionId,
        transactionParams: {
          sender: {
            id: '',
            address: '',
          },
          sendAsset: {
            amount: '0',
            assetDetails: {
              type: AssetType.native,
              image: './images/placeholder.svg',
              symbol: '',
              balance: '0',
              details: {
                decimals: 8,
              },
            },
            asset: '',
            denominatinon: undefined,
            valid: false,
            error: '',
          },
          receiveAsset: {
            amount: '0',
            asset: '',
            denominatinon: undefined,
            valid: false,
            error: '',
          },
          recipient: {
            address: '',
            valid: false,
            error: '',
          },
          fee: {
            valid: false,
            error: '',
            isLoading: true,
            fee: '0',
            unit: '',
            confirmationTime: '',
            feeInFiat: '',
            feeLevel: FeeLevel.Average,
          },
          data: {},
          network: {
            network: MultichainNetworks.BITCOIN,
            error: '',
          },
        },
        transaction: null,
        valid: false,
        error: '',
      },
    },
    currentTransactionUUID: mockTransactionId,
    error: '',
  },
};

const createCombinedStateWithUser = (
  multichainSendState: CombinedBackgroundAndReduxState,
  account?: InternalAccount,
): Partial<CombinedBackgroundAndReduxState> => {
  return {
    ...multichainSendState,
    metamask: {
      ...multichainSendState.metamask,
      internalAccounts: {
        ...multichainSendState.metamask.internalAccounts,
        accounts: {
          ...multichainSendState.metamask.internalAccounts.accounts,
          // conditionally add account here if its defined
          ...(account ? { [account.id]: account } : {}),
        },
        selectedAccount: account
          ? account.id
          : multichainSendState.metamask.internalAccounts.selectedAccount,
      },
    },
    multichainSend: multichainSendState.multichainSend,
  };
};

// This is only used to test async actions that has side effects
// This store only contains the multichainSend slice
// TODO: fix combined reducer state type
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const createRealStore = (state: any) => {
  const actions: AnyAction[] = [];
  const listenerMiddleware = createListenerMiddleware();

  listenerMiddleware.startListening({
    predicate: () => true, // Listen to all actions
    effect: (action) => {
      actions.push(action);
    },
  });
  const store = configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware().prepend(listenerMiddleware.middleware),
    preloadedState: state,
  });

  return {
    actions,
    store,
  };
};

// TODO: fix any
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const render = async (state: any, account?: InternalAccount) => {
  const { store, actions } = createRealStore(
    createCombinedStateWithUser(state, account),
  );

  let result;

  await act(
    // @ts-expect-error TODO: fix type error
    async () => (result = renderWithProvider(<MultichainSendPage />, store)),
  );

  return { store, result: result as unknown as RenderResult, actions };
};

describe('MultichainSendPage', () => {
  it('renders correctly', async () => {
    const {
      result: { getByText, container },
    } = await render(baseStore);

    expect(getByText('Send')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('creates a new draft transaction if one does not exist', async () => {
    const mockStoreWithNoDrafts = {
      ...baseStore,
      multichainSend: {
        ...initialMultichainSendState,
      },
    };

    const { actions } = await render(mockStoreWithNoDrafts);
    const selectedAccount = getSelectedInternalAccount(mockStoreWithNoDrafts);
    // TODO: fix type error
    // @ts-expect-error selector type should be the same a combined reducer state
    const network = getMultichainProviderConfig(mockStoreWithNoDrafts).chainId;
    expect(actions[1]).toStrictEqual({
      type: addNewDraft.type,
      payload: {
        account: selectedAccount,
        network,
      },
    });
  });

  describe('Account selector', () => {
    it('only shows non EVM account', async () => {
      const {
        result: { getByTestId, getAllByTestId },
      } = await render(baseStore);

      const accountSelector = getByTestId('send-page-account-picker');

      fireEvent.click(accountSelector);

      expect(getAllByTestId('account-list-item-menu-button')).toHaveLength(1);
    });
  });

  describe('Recipient input', () => {
    const mockRecipient =
      'bc1qwzrryqr3ja8w7hnja2spmkgfdcgvqwp5swz4af4ngsjecfz0w0pqud7k38';
    const mockInvalidRecipient = 'mock';
    const selectedAccount = getSelectedInternalAccount(baseStore);
    // TODO: fix type error
    // @ts-expect-error selector type should be the same a combined reducer state
    const network = getMultichainProviderConfig(baseStore).chainId;

    it("should update the recipient's address", async () => {
      const {
        actions,
        result: { getByTestId },
      } = await render(baseStore);

      const input = getByTestId('multichain-send-recipient-input');

      fireEvent.change(input, { target: { value: mockRecipient } });

      await waitFor(() => {
        expect(actions[1]).toStrictEqual({
          type: addNewDraft.type,
          payload: {
            account: selectedAccount,
            network,
          },
        });
      });
    });

    it("should clear the recipient's address", async () => {
      const {
        actions,
        result: { getByTestId, getByLabelText },
      } = await render(baseStoreWithDraft);

      const input = getByTestId('multichain-send-recipient-input');

      fireEvent.change(input, { target: { value: mockRecipient } });

      await waitFor(() => {
        expect(actions[1]).toStrictEqual({
          type: 'multichainSend/updateAndValidateRecipient/fulfilled',
          meta: expect.any(Object),
          payload: {
            address: mockRecipient,
            error: '',
            valid: true,
          },
        });
      });

      const recipient = getByTestId('multichain-recipient');
      expect(recipient).toHaveTextContent(shortenAddress(mockRecipient));

      const clearButton = getByLabelText('Close');
      fireEvent.click(clearButton);

      await waitFor(() => {
        expect(actions[2]).toStrictEqual({
          type: editTransaction.type,
          payload: {
            recipient: { address: '', valid: false },
          },
        });
      });
    });

    it("shows an error message if the recipient's address is invalid", async () => {
      const {
        actions,
        result: { getByTestId },
      } = await render(baseStoreWithDraft);

      const input = getByTestId('multichain-send-recipient-input');

      fireEvent.change(input, { target: { value: mockInvalidRecipient } });

      await waitFor(() => {
        expect(actions[1]).toStrictEqual({
          type: 'multichainSend/updateAndValidateRecipient/rejected',
          meta: expect.any(Object),
          payload: undefined,
          error: expect.any(Object),
        });
        const recipientError = getByTestId('multichain-recipient-error');
        expect(recipientError).toHaveTextContent(
          messages.invalidAddressRecipient.message,
        );
      });
    });
  });

  describe('Amount Input', () => {
    const mockAmount = 0.00123;
    const mockAmountInHex = '0x1e078';
    const mockInvalidAmount = 'invalid';
    const mockLargeAmount = 100;
    const mockLargeAmountInHex = '0x2540be400';
    const mockZeroAmount = 0;
    const mockZeroHexAmount = '0x0';

    it('updates the amount and estimates fees', async () => {
      const {
        actions,
        result: { getByTestId },
      } = await render(baseStoreWithDraft);
      const selectedAccount = getSelectedInternalAccount(baseStoreWithDraft);
      // @ts-expect-error this builder is mocked, therefore no need for constructor args
      const expectedFees = await new MockTransactionBuilder().estimateGas();

      const input = getByTestId('currency-input');

      fireEvent.change(input, { target: { value: mockAmount } });

      await waitFor(() => {
        expect(actions[0]).toStrictEqual({
          type: updateSendAmount.type,
          payload: mockAmountInHex, // the amount input converts the user input value to hex
        });

        expect(actions[1]).toStrictEqual({
          type: 'multichainSend/estimateFee/pending',
          payload: undefined,
          meta: {
            arg: {
              account: selectedAccount,
              transactionId: mockTransactionId,
            },
            requestId: expect.any(String),
            requestStatus: 'pending',
          },
        });

        expect(actions[2]).toStrictEqual({
          type: 'multichainSend/estimateFee/fulfilled',
          meta: expect.any(Object),
          payload: expectedFees,
        });

        expect(getByTestId('currency-input')).toHaveValue(mockAmount);
        expect(getByTestId('multichain-confirmation-fee')).toHaveTextContent(
          expectedFees.fee,
        );
      });
    });

    // @ts-expect-error This is missing from the Mocha type definitions
    it.each([
      {
        amount: mockLargeAmount,
        amountInHex: mockLargeAmountInHex,
        error: messages.insufficientFunds.message,
      },
      {
        amount: mockZeroAmount,
        amountInHex: mockZeroHexAmount,
        error: messages.negativeOrZeroAmountToken.message,
      },
    ])(
      `shows error '$error' for amount $amount`,
      async ({
        amount,
        amountInHex,
        error,
      }: {
        amount: number;
        amountInHex: string;
        error: string;
      }) => {
        const {
          actions,
          result: { getByTestId },
        } = await render(baseStoreWithDraft);

        const input = getByTestId('currency-input');

        fireEvent.change(input, { target: { value: amount } });

        await waitFor(() => {
          expect(actions[0]).toStrictEqual({
            type: updateSendAmount.type,
            payload: amountInHex, // the amount input converts the user input value to hex
          });
          expect(getByTestId('send-page-amount-error')).toHaveTextContent(
            error,
          );
        });
      },
    );

    it("shows an error message if the amount's value is invalid", async () => {
      const {
        actions,
        result: { getByTestId },
      } = await render(baseStoreWithDraft);

      const input = getByTestId('currency-input');

      fireEvent.change(input, { target: { value: mockInvalidAmount } });

      await waitFor(() => {
        expect(actions).toHaveLength(0);
      });
    });
  });

  describe('Footer buttons', () => {
    afterEach(() => {
      jest.resetAllMocks();
    });

    describe('Cancel button', () => {
      it('clears the draft transaction on cancel', async () => {
        const {
          actions,
          result: { getByText },
        } = await render(baseStoreWithDraft);

        const cancelButton = getByText(messages.cancel.message);
        fireEvent.click(cancelButton);

        expect(actions[0]).toStrictEqual({
          type: 'multichainSend/clearDraft',
          payload: undefined,
        });

        expect(mockHistoryPush).toHaveBeenCalledWith(DEFAULT_ROUTE);
      });
    });

    describe('Confirm button', () => {
      it('updates the stage to pending confirmation and redirects to the confirmation route', async () => {
        const {
          actions,
          result: { getByText },
        } = await render({
          ...baseStoreWithDraft,
          multichainSend: {
            ...baseStoreWithDraft.multichainSend,
            draftTransactions: {
              [baseStoreWithDraft.multichainSend.currentTransactionUUID]: {
                ...baseStoreWithDraft.multichainSend.draftTransactions[
                  baseStoreWithDraft.multichainSend
                    .currentTransactionUUID as keyof typeof baseStoreWithDraft.multichainSend.draftTransactions
                ],
                valid: true,
                transactionParams: {
                  ...baseStoreWithDraft.multichainSend.draftTransactions[
                    baseStoreWithDraft.multichainSend
                      .currentTransactionUUID as keyof typeof baseStoreWithDraft.multichainSend.draftTransactions
                  ].transactionParams,
                  recipient: {
                    address:
                      'bc1qwzrryqr3ja8w7hnja2spmkgfdcgvqwp5swz4af4ngsjecfz0w0pqud7k38',
                    valid: true,
                    error: '',
                  },
                  sendAsset: {
                    amount: '0.000123',
                    assetDetails: {
                      type: AssetType.native,
                      image: './images/placeholder.svg',
                      symbol: '',
                      balance: '0',
                      details: {
                        decimals: 8,
                      },
                    },
                    asset: '',
                    denominatinon: undefined,
                    valid: true,
                    error: '',
                  },
                  receiveAsset: {
                    amount: '0',
                    asset: '',
                    denominatinon: undefined,
                    valid: true,
                    error: '',
                  },
                  fee: {
                    valid: true,
                    error: '',
                    isLoading: false,
                    fee: '0.000123',
                    unit: 'BTC',
                    confirmationTime: '10 minutes',
                    feeInFiat: '0',
                    feeLevel: FeeLevel.Average,
                  },
                  data: {},
                  network: {
                    network: MultichainNetworks.BITCOIN,
                    error: '',
                  },
                },
              },
            },
          },
        });

        const reviewButton = getByText(messages.review.message);
        fireEvent.click(reviewButton);

        await waitFor(() => {
          expect(actions[0]).toStrictEqual({
            type: updateStage.type,
            payload: { stage: SendStage.PENDING_CONFIRMATION },
          });
          expect(mockHistoryPush).toHaveBeenCalledWith(
            `${MULTICHAIN_CONFIRM_TRANSACTION_ROUTE}/test-id`,
          );
        });
      });

      it('review is disabled if the transaction is invalid', async () => {
        const {
          result: { getByText },
        } = await render(baseStore);

        const reviewButton = getByText(messages.review.message);
        expect(reviewButton).toBeDisabled();
        expect(mockHistoryPush).not.toHaveBeenCalled();
      });
    });
  });
});
