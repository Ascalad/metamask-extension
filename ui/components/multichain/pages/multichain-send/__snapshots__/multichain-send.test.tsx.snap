// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultichainSendPage renders correctly 1`] = `
<div>
  <div
    class="mm-box multichain-page mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-center mm-box--width-full mm-box--height-full mm-box--background-color-background-alternative"
  >
    <div
      class="mm-box multichain-page__inner-container multichain-send-page mm-box--display-flex mm-box--flex-direction-column mm-box--width-full mm-box--height-full mm-box--background-color-background-alternative"
    >
      <div
        class="mm-box mm-header-base multichain-page-header mm-box--padding-4 mm-box--display-flex mm-box--justify-content-center mm-box--width-full"
      >
        <div
          class="mm-box"
          style="min-width: 0px;"
        >
          <button
            aria-label="Back"
            class="mm-box mm-button-icon mm-button-icon--size-sm mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-icon-default mm-box--background-color-transparent mm-box--rounded-lg"
          >
            <span
              class="mm-box mm-icon mm-icon--size-sm mm-box--display-inline-block mm-box--color-inherit"
              style="mask-image: url('./images/icons/arrow-left.svg');"
            />
          </button>
        </div>
        <div
          class="mm-box"
        >
          <h4
            class="mm-box mm-text mm-text--heading-sm mm-text--ellipsis mm-text--text-align-center mm-box--padding-inline-start-8 mm-box--padding-inline-end-8 mm-box--display-block mm-box--color-text-default"
          >
            Send
          </h4>
        </div>
      </div>
      <div
        class="mm-box multichain-page-content mm-box--padding-4 mm-box--display-flex mm-box--flex-direction-column mm-box--width-full mm-box--height-full mm-box--background-color-background-alternative"
      >
        <div
          class="mm-box mm-box--padding-bottom-4 mm-box--display-flex mm-box--flex-direction-column"
        >
          <label
            class="mm-box mm-text mm-label mm-text--body-md mm-text--font-weight-medium mm-box--padding-bottom-2 mm-box--display-inline-flex mm-box--align-items-center mm-box--color-text-default"
          >
            From
          </label>
          <button
            class="mm-box mm-text mm-button-base mm-button-base--size-lg mm-button-base--block mm-button-base--ellipsis multichain-account-picker multichain-send-page__account-picker mm-text--body-md-medium mm-text--ellipsis mm-box--padding-0 mm-box--padding-top-4 mm-box--padding-right-4 mm-box--padding-bottom-4 mm-box--padding-left-3 mm-box--display-inline-flex mm-box--gap-2 mm-box--justify-content-flex-start mm-box--align-items-center mm-box--width-full mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-lg mm-box--border-color-border-muted mm-box--border-width-1 box--border-style-solid"
            data-testid="send-page-account-picker"
          >
            <span
              class="mm-box mm-text mm-text--inherit mm-text--ellipsis mm-box--display-flex mm-box--gap-2 mm-box--align-items-center mm-box--width-full mm-box--color-text-default"
            >
              <div
                class="mm-box mm-text mm-avatar-base mm-avatar-base--size-md mm-avatar-account mm-text--body-sm mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-background-default box--border-style-solid box--border-width-1"
              >
                <div
                  class="mm-avatar-account__jazzicon"
                >
                  <div
                    style="border-radius: 50px; overflow: hidden; padding: 0px; margin: 0px; width: 32px; height: 32px; display: inline-block; background: rgb(24, 162, 242);"
                  >
                    <svg
                      height="32"
                      width="32"
                      x="0"
                      y="0"
                    >
                      <rect
                        fill="#F29602"
                        height="32"
                        transform="translate(0.001112151990700775 -0.0005016734084641463) rotate(358.8 16 16)"
                        width="32"
                        x="0"
                        y="0"
                      />
                      <rect
                        fill="#FA6C00"
                        height="32"
                        transform="translate(7.965719514969553 10.506673824525246) rotate(69.5 16 16)"
                        width="32"
                        x="0"
                        y="0"
                      />
                      <rect
                        fill="#236CE1"
                        height="32"
                        transform="translate(-19.066706584002095 16.199592375372838) rotate(260.2 16 16)"
                        width="32"
                        x="0"
                        y="0"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <div
                class="mm-box mm-text mm-avatar-base mm-avatar-base--size-md mm-avatar-account mm-text--body-sm mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-background-default box--border-style-solid box--border-width-1"
              >
                <div
                  class="mm-avatar-account__jazzicon"
                >
                  <div
                    style="border-radius: 50px; overflow: hidden; padding: 0px; margin: 0px; width: 32px; height: 32px; display: inline-block; background: rgb(24, 162, 242);"
                  >
                    <svg
                      height="32"
                      width="32"
                      x="0"
                      y="0"
                    >
                      <rect
                        fill="#F29602"
                        height="32"
                        transform="translate(0.001112151990700775 -0.0005016734084641463) rotate(358.8 16 16)"
                        width="32"
                        x="0"
                        y="0"
                      />
                      <rect
                        fill="#FA6C00"
                        height="32"
                        transform="translate(7.965719514969553 10.506673824525246) rotate(69.5 16 16)"
                        width="32"
                        x="0"
                        y="0"
                      />
                      <rect
                        fill="#236CE1"
                        height="32"
                        transform="translate(-19.066706584002095 16.199592375372838) rotate(260.2 16 16)"
                        width="32"
                        x="0"
                        y="0"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <span
                class="mm-box mm-text multichain-account-picker__label multichain-send-page__account-picker__label mm-text--body-md mm-text--ellipsis mm-box--padding-inline-start-1 mm-box--color-text-default"
                style="flex-grow: 1; text-align: start;"
              >
                Bitcoin Account
                <p
                  class="mm-box mm-text mm-text--body-sm mm-text--ellipsis mm-text--text-align-start mm-box--display-flex mm-box--color-text-alternative"
                >
                  bc1qa4m...n7c6x
                </p>
              </span>
            </span>
            <span
              class="mm-box mm-icon mm-icon--size-sm mm-box--margin-inline-start-1 mm-box--display-inline-block mm-box--color-icon-default"
              style="mask-image: url('./images/icons/arrow-down.svg');"
            />
          </button>
        </div>
        <div
          class="mm-box asset-picker-amount"
        >
          <div
            class="mm-box mm-box--margin-bottom-1 mm-box--padding-1 mm-box--padding-top-1 mm-box--padding-bottom-1 mm-box--display-flex mm-box--align-items-center mm-box--background-color-background-default mm-box--rounded-lg mm-box--border-style-solid mm-box--border-color-border-muted mm-box--border-width-1"
          >
            <button
              class="mm-box mm-text mm-button-base mm-button-base--size-md mm-button-base--disabled asset-picker mm-button-primary mm-button-primary--disabled mm-text--body-md-medium mm-box--padding-2 mm-box--padding-right-2 mm-box--padding-left-2 mm-box--display-flex mm-box--gap-2 mm-box--align-items-center mm-box--color-primary-inverse mm-box--background-color-transparent mm-box--rounded-pill"
              data-testid="asset-picker-button"
              data-theme="light"
              disabled=""
            >
              <span
                class="mm-box mm-text mm-text--inherit mm-box--color-primary-inverse"
              >
                <div
                  class="mm-box mm-box--display-flex mm-box--gap-3 mm-box--align-items-center"
                >
                  <div
                    class="mm-box mm-box--display-flex"
                  >
                    <div
                      class="mm-box mm-badge-wrapper mm-box--display-inline-block"
                    >
                      <div
                        class="mm-box mm-text mm-avatar-base mm-avatar-base--size-md mm-avatar-token mm-text--body-sm mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full"
                      >
                        m
                      </div>
                      <div
                        class="mm-box mm-badge-wrapper__badge-container mm-badge-wrapper__badge-container--circular-top-right"
                      >
                        <div
                          class="mm-box mm-text mm-avatar-base mm-avatar-base--size-xs mm-avatar-network mm-text--body-xs mm-text--text-transform-uppercase mm-box--display-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-text-default mm-box--background-color-background-alternative mm-box--rounded-full mm-box--border-color-border-default box--border-style-solid box--border-width-1"
                        >
                          <img
                            alt="Bitcoin logo"
                            class="mm-avatar-network__network-image"
                            src="./images/bitcoin-logo.svg"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                  <div>
                    <div
                      aria-describedby="tippy-tooltip-2"
                      class=""
                      data-original-title="mock-ticker"
                      data-tooltipped=""
                      style="display: inline;"
                      tabindex="0"
                    >
                      <p
                        class="mm-box mm-text asset-picker__symbol mm-text--body-md mm-box--color-text-default"
                      >
                        mock-t...
                      </p>
                    </div>
                  </div>
                </div>
              </span>
            </button>
            <div
              class="unit-input asset-picker-amount__input"
            >
              <div
                class="unit-input__inputs"
              >
                <div
                  class="unit-input__input-container"
                  style="display: inherit;"
                  title=""
                >
                  <input
                    class="unit-input__input"
                    data-testid="currency-input"
                    dir="ltr"
                    min="0"
                    placeholder="0"
                    step="any"
                    style="width: 1.5ch;"
                    type="number"
                    value=""
                  />
                  <div
                    class="unit-input__suffix"
                  >
                    BTC
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            class="mm-box mm-box--display-flex"
          >
            <div
              class="mm-box asset-picker-amount__balance mm-box--display-flex"
            >
              <p
                class="mm-box mm-text mm-text--body-sm mm-box--margin-right-1 mm-box--color-text-alternative"
              >
                Balance
                :
              </p>
              <div
                class="mm-box currency-display-component mm-box--display-flex mm-box--flex-wrap-wrap mm-box--align-items-center"
                title="222222"
              >
                <span
                  class="mm-box mm-text currency-display-component__text mm-text--body-sm mm-text--ellipsis mm-box--color-text-alternative"
                >
                  222222
                </span>
              </div>
            </div>
            <button
              class="mm-box mm-text mm-button-base asset-picker-amount__max-clear mm-button-link mm-button-link--size-auto mm-text--body-md-medium mm-box--margin-left-auto mm-box--padding-0 mm-box--padding-right-0 mm-box--padding-left-0 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-primary-default mm-box--background-color-transparent"
              data-testid="max-clear-button"
            >
              Max
            </button>
          </div>
        </div>
        <div
          class="mm-box mm-box--margin-top-6"
        >
          <div
            class="mm-box mm-box--padding-bottom-4 mm-box--display-flex mm-box--flex-direction-column"
          >
            <label
              class="mm-box mm-text mm-label mm-text--body-md mm-text--font-weight-medium mm-box--padding-bottom-2 mm-box--display-inline-flex mm-box--align-items-center mm-box--color-text-default"
            >
              To
            </label>
            <div
              class="mm-box mm-box--padding-4 mm-box--sm:padding-3 mm-box--md:padding-4 mm-box--lg:padding-3 mm-box--display-flex mm-box--align-items-center mm-box--background-color-background-default mm-box--rounded-md mm-box--border-style-solid mm-box--border-color-border-muted box--border-width-1"
            >
              <div
                class="mm-box mm-box--display-flex mm-box--flex-direction-column mm-box--width-full"
              >
                <div
                  class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-space-between mm-box--width-full"
                >
                  <input
                    autocomplete="off"
                    class="mm-box mm-text mm-input mm-text--body-md mm-box--margin-0 mm-box--padding-0 mm-box--display-block mm-box--width-full mm-box--color-text-default mm-box--background-color-transparent mm-box--border-style-none"
                    data-testid="multichain-send-recipient-input"
                    placeholder="Enter a valid Bitcoin address"
                    spellcheck="false"
                    type="text"
                    value=""
                  />
                  
                </div>
              </div>
            </div>
            
          </div>
        </div>
        <div
          class="mm-box mm-box--margin-bottom-2 mm-box--padding-4 mm-box--gap-4 mm-box--flex-direction-column mm-box--justify-content-center mm-box--align-items-center mm-box--background-color-background-default mm-box--rounded-lg"
        >
          <div
            class="mm-box mm-box--margin-bottom-2 mm-box--display-flex mm-box--flex-direction-row mm-box--justify-content-space-between"
          >
            <div
              class="mm-box mm-box--display-flex mm-box--flex-direction-row"
            >
              <div>
                <div
                  aria-describedby="tippy-tooltip-1"
                  class=""
                  data-original-title="The fee is an estimate and may vary based on network conditions."
                  data-tooltipped=""
                  style="display: inline;"
                  tabindex="0"
                >
                  <div
                    class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--align-items-center"
                  >
                    <p
                      class="mm-box mm-text mm-text--body-md mm-text--font-weight-medium mm-box--color-text-default"
                    >
                      Estimated Fee
                    </p>
                    <span
                      class="mm-box mm-icon mm-icon--size-sm mm-box--margin-left-1 mm-box--display-inline-block mm-box--color-icon-muted"
                      style="mask-image: url('./images/icons/question.svg');"
                    />
                  </div>
                </div>
              </div>
            </div>
            <div
              class="mm-box loading-item mm-box--display-flex mm-box--flex-direction-row"
            >
              Loading
            </div>
          </div>
          
        </div>
        <div
          class="mm-box mm-box--display-flex mm-box--flex-direction-row mm-box--align-items-center"
        >
          <span
            class="mm-box mm-icon mm-icon--size-sm mm-box--margin-right-1 mm-box--display-inline-block mm-box--color-icon-muted"
            style="mask-image: url('./images/icons/info.svg');"
          />
          <p
            class="mm-box mm-text mm-text--body-md mm-box--color-text-alternative"
          >
            Sat protection included
          </p>
        </div>
      </div>
      <div
        class="mm-box multichain-page-footer mm-box--padding-4 mm-box--display-flex mm-box--gap-4 mm-box--width-full mm-box--border-style-none"
      >
        <button
          class="mm-box mm-text mm-button-base mm-button-base--size-lg mm-button-base--block mm-button-secondary mm-text--body-md-medium mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-4 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-primary-default mm-box--background-color-transparent mm-box--rounded-pill mm-box--border-color-primary-default box--border-style-solid box--border-width-1"
        >
          Cancel
        </button>
        <button
          class="mm-box mm-text mm-button-base mm-button-base--size-lg mm-button-base--disabled mm-button-base--block mm-button-primary mm-button-primary--disabled mm-text--body-md-medium mm-box--padding-0 mm-box--padding-right-4 mm-box--padding-left-4 mm-box--display-inline-flex mm-box--justify-content-center mm-box--align-items-center mm-box--color-primary-inverse mm-box--background-color-primary-default mm-box--rounded-pill"
          data-theme="light"
          disabled=""
        >
          Review
        </button>
      </div>
    </div>
  </div>
</div>
`;
